File: benchmark/foreign/R/biogeme_swissmetro.R
# Run the Biogeme Swissmetro example in R

library(apollo)
library(tictoc)

apollo_initialise()

apollo_control = list(
  modelName = "biogeme_swissmetro",
  modelDescr = "Biogeme Swissmetro example",
  indivID = "ID",
  outputDirectory = tempdir()
)

# can't find a way to do source-relative paths in R
database = read.delim("../../../test/data/biogeme_swissmetro.dat")
database = subset(database, (PURPOSE == 1 | PURPOSE == 3) & CHOICE != 0)

database = within(database, {
  SM_COST = SM_CO * (GA == 0)
  TRAIN_COST = TRAIN_CO * (GA == 0)
  CAR_AV_SP = CAR_AV * (SP != 0)
  TRAIN_AV_SP = TRAIN_AV* (SP != 0)
  TRAIN_TT_SCALED = TRAIN_TT / 100
  TRAIN_COST_SCALED = TRAIN_COST / 100
  SM_TT_SCALED = SM_TT / 100
  SM_COST_SCALED = SM_COST / 100
  CAR_TT_SCALED = CAR_TT / 100
  CAR_CO_SCALED = CAR_CO / 100
})

# data loading and prep done, time the model setup and run
tic()

apollo_beta = c(
  asc_car = 0,
  asc_train = 0, 
  asc_sm = 0,
  b_time = 0,
  b_cost = 0
)

apollo_fixed = c("asc_sm")

apollo_inputs = apollo_validateInputs()

apollo_probabilities = function (apollo_beta, apollo_inputs, functionality="estimate") {
  apollo_attach(apollo_beta, apollo_inputs)
  on.exit(apollo_detach(apollo_beta, apollo_inputs))
  
  # probabilities
  P = list()
  
  # utilities
  V = list(
    train = asc_train + b_time * TRAIN_TT_SCALED + b_cost * TRAIN_COST_SCALED,
    Swissmetro = asc_sm + b_time * SM_TT_SCALED + b_cost * SM_COST_SCALED,
    car = asc_car + b_time * CAR_TT_SCALED + b_cost * CAR_CO_SCALED
  )
  
  mnl_settings = list(
    alternatives=c(train = 1, Swissmetro = 2, car = 3),
    avail = list(train = TRAIN_AV_SP, Swissmetro = SM_AV, car = CAR_AV_SP),
    choiceVar = CHOICE,
    utilities = V
  )
  
  P[["model"]] = apollo_mnl(mnl_settings, functionality)
  
  # panel multiplications should not matter in an MNL but it is in the Apollo examples
  P = apollo_panelProd(P, apollo_inputs, functionality)
  
  P = apollo_prepareProb(P, apollo_inputs, functionality)
  
  return(P)
}

model = apollo_estimate(apollo_beta, apollo_fixed, apollo_probabilities, apollo_inputs)

timer = toc()

print(paste("Model estimation time:", (timer$toc - timer$tic)[[1]], "seconds"))

File: benchmark/foreign/python/biogeme_swissmetro.py
# Biogeme's Swissmetro example: https://github.com/michelbierlaire/biogeme/blob/master/examples/swissmetro/01logit.ipynb
# this code is copied and lightly modified from the Biogeme example, which is under the Biogeme open-source license

import pandas as pd
import biogeme.database as db
import biogeme.biogeme as bio
from biogeme import models
from biogeme.expressions import Beta
from benchmarkable import Benchmarkable
import os


class BiogemeSwissmetro(Benchmarkable):
    def setup(self):
        # Read the data
        df = pd.read_csv(
            os.path.join(
                os.path.dirname(__file__), "../../../test/data/biogeme_swissmetro.dat"
            ),
            sep="\t",
        )
        self.database = db.Database("swissmetro", df)

        self.dv = dv = self.database.variables

        # Removing some observations
        exclude = (
            (dv["PURPOSE"] != 1) * (dv["PURPOSE"] != 3) + (dv["CHOICE"] == 0)
        ) > 0
        self.database.remove(exclude)

        # Definition of new variables
        self.SM_COST = dv["SM_CO"] * (dv["GA"] == 0)
        self.TRAIN_COST = dv["TRAIN_CO"] * (dv["GA"] == 0)
        self.CAR_AV_SP = dv["CAR_AV"] * (dv["SP"] != 0)
        self.TRAIN_AV_SP = dv["TRAIN_AV"] * (dv["SP"] != 0)
        self.TRAIN_TT_SCALED = dv["TRAIN_TT"] / 100
        self.TRAIN_COST_SCALED = self.TRAIN_COST / 100
        self.SM_TT_SCALED = dv["SM_TT"] / 100
        self.SM_COST_SCALED = self.SM_COST / 100
        self.CAR_TT_SCALED = dv["CAR_TT"] / 100
        self.CAR_CO_SCALED = dv["CAR_CO"] / 100

    def measurable(self):
        # for comparability with DCM.jl, do all model setup here, but create 
        # derived variables outside the timed loop
        # Parameters to be estimated
        ASC_CAR = Beta("ASC_CAR", 0, None, None, 0)
        ASC_TRAIN = Beta("ASC_TRAIN", 0, None, None, 0)
        ASC_SM = Beta("ASC_SM", 0, None, None, 1)
        B_TIME = Beta("B_TIME", 0, None, None, 0)
        B_COST = Beta("B_COST", 0, None, None, 0)

        # Definition of the utility functions
        V1 = ASC_TRAIN + B_TIME * self.TRAIN_TT_SCALED + B_COST * self.TRAIN_COST_SCALED
        V2 = ASC_SM + B_TIME * self.SM_TT_SCALED + B_COST * self.SM_COST_SCALED
        V3 = ASC_CAR + B_TIME * self.CAR_TT_SCALED + B_COST * self.CAR_CO_SCALED

        # Associate utility functions with the numbering of alternatives
        V = {1: V1, 2: V2, 3: V3}

        # Associate the availability conditions with the alternatives
        av = {1: self.TRAIN_AV_SP, 2: self.dv["SM_AV"], 3: self.CAR_AV_SP}

        # Definition of the model. This is the contribution of each
        # observation to the log likelihood function.
        logprob = models.loglogit(V, av, self.dv["CHOICE"])

        # Create the Biogeme object
        self.biogeme = bio.BIOGEME(self.database, logprob)
        self.biogeme.modelName = "01logit"

        # actually do the estimation
        self.biogeme.estimate()
